src/library/scala/collection/mutable/ListBuffer.scala

```diff
@@ -321,12 +358,24 @@ class ListBuffer[A]
   }
 
   def patchInPlace(from: Int, patch: collection.IterableOnce[A], replaced: Int): this.type = {
-    val i = math.min(math.max(from, 0), length)
-    val n = math.min(math.max(replaced, 0), length)
-    ensureUnaliased()
-    val p = locate(i)
-    removeAfter(p, math.min(n, len - i))
-    insertAfter(p, patch.iterator)
+    val _len = len
+    val _from = math.max(from, 0)         // normalized
+    val _replaced = math.max(replaced, 0) // normalized
+    val it = patch.iterator
+
+    val nonEmptyPatch = it.hasNext
+    val nonEmptyReplace = (_from < _len) && (_replaced > 0)
+
+    // don't want to add a mutation or check aliasing (potentially expensive)
+    // if there's no patching to do
+    if (nonEmptyPatch || nonEmptyReplace) {
+      ensureUnaliased()
+      val i = math.min(_from, _len)
+      val n = math.min(_replaced, _len)
+      val p = locate(i)
+      removeAfter(p, math.min(n, _len - i))
+      insertAfter(p, it)
```

Comment:
this fails with a `ConcurrentModificationException` if `it` comes from `this`. might fix in a separate PR. it didn't work before this PR anyway - it failed in an unspecified way (adding/removing the wrong elements) rather than with an exception

Reply:
specifically, given the following code

```scala
val b = ListBuffer(1, 2, 3)
b.patchInPlace(1, b, 1)
```

would previously result in `ListBuffer(1, 1, 3, 3)` when it should result in `ListBuffer(1, 1, 2, 3, 3)`, but now throws an exception